src/plan_expand_hypertable.c

```diff
@@ -312,6 +316,138 @@ get_chunk_oids(CollectQualCtx *ctx, PlannerInfo *root, RelOptInfo *rel, Hypertab
 		return get_explicit_chunk_oids(ctx, ht);
 }
 
+#if !(PG96 || PG10)
+
+/*
+ * Check whether a group clause uses a function on an open dimension.
+ *
+ * An open dimension can be partitioned using a function expression. In that
+ * case, try to support partitionwise aggregation both on the raw column and the
+ * function expression. For instance, a query could look like
+ *
+ * SELECT time, device, avg(temp)
+ * FROM hypertable
+ * GROUP BY 1, 2;
+ *
+ * or,
+ *
+ * SELECT func(time), device, avg(temp)
+ * FROM hypertable
+ * GROUP BY 1, 2;
+ *
+ * We can support partitionwise aggregation on the first query, but also on the
+ * second if 'func' is the partitioning function on the open "time" dimension.
+ *
+ * Thus, this function will return TRUE, iff either one of these two cases
+ * occur.
+ */
+static bool
+open_dim_uses_partfunc(List **groupexprs, Query *parse, Dimension *dim)
+{
+	if (!IS_OPEN_DIMENSION(dim) || NULL == dim->partitioning)
+		return false;
+
+	/* Lazy set groupexpr */
+	if (*groupexprs == NIL)
+		*groupexprs = get_sortgrouplist_exprs(parse->groupClause, parse->targetList);
+
+	return list_member(*groupexprs, dim->partitioning->partfunc.func_fmgr.fn_expr);
+}
+
+/*
+ * Create partition expressions for a hypertable.
+ *
+ * Build an array of partition expressions where each element is a list of
+ * expression for a particular partitioning dimension.
+ *
+ * The partition expressions are used by, e.g., group_by_has_partkey() to check
+ * whether a GROUP BY clause covers all partitioning dimesions.
+ */
+static List **
+get_hypertable_partexprs(Hypertable *ht, Query *parse, Index varno)
+{
+	int i;
+	List **partexprs;
+	List *groupexprs = NIL;
+
+	Assert(NULL != ht->space);
+
+	partexprs = palloc0(sizeof(List *) * ht->space->num_dimensions);
+
+	for (i = 0; i < ht->space->num_dimensions; i++)
+	{
+		Dimension *dim = &ht->space->dimensions[i];
+		Expr *expr;
+
+		/* Only functions on open dimensions need to be function expressions */
+		if (open_dim_uses_partfunc(&groupexprs, parse, dim))
+			expr = (Expr *) dim->partitioning->partfunc.func_fmgr.fn_expr;
```

Comment:
I am a bit confused why we don't always add the expression for the partitioning function. my read is `group_by_has_partkey` will return true if there is a match for any list element so why don't we always add a var and a func expr?

Reply:
I don't think that's the case. `group_by_has_partkey` will only return true if _all_ partkeys exists in the group by clause. So, if we add both the original column and the function expression, then there will never be a match. Therefore, we need to add either the "raw" column or the expression on the column. See comment on `group_by_has_partkey`:

```
* group_by_has_partkey
 *
 * Returns true, if all the partition keys of the given relation are part of
 * the GROUP BY clauses, false otherwise.
```

Reply:
So I think there is confusion here. rel->partexpr is an array of lists. Each list in the array represents one partition key. But the list itself is pretty much an OR of exprs. if you look at group_by_has_partkey it checks that any one of the list elements matches a group by clause for every element in the array. So each list can contain both exprs and everything should work fine